Implement Byte-Aware Chunking

dezren39 opened the pull request:
- Add chunkType parameter, allowing multi-byte aware chunking.
   - Third parameter, default behavior if excluded is equal to previous behavior.
   - "-1"  : same as previous default behaviour. every character = 1
   - "0"   : every character = as many bytes as the character is
   - ">0" : every character = as many bytes as the character is, up to N
     - This is the use case I need, as I need each emoji to count as "2".
- Some characters may encode to more bytes than one might have expected.
- Added new tests to accommodate changes and verify backwards compatibility.
     - When working with emoji, I've had issues where they display differently everywhere.
     - Parsing the variable names has been easier overall, while sometimes more annoying on properly encoding editors.
- Changed chunkSize validation to use Numbers.parseInt and Numbers.isNaN
  - It has better overall compatibility, though does allow weird things like `chunk("text", "2 years", "-1 years")`
- Updated dependencies.
   - Added `fast-text-encoding` in order to fill `TextEncoder` in `jest`
- Incremented minor version, because although adding a parameter isn't a breaking change --
   - I do worry the overall changes may have some possibility of compatibility issue. Babel helps alleviate this concern though.
   - The effort to allow this capability has increased the complexity of this project.
      - I've tried to add enough testing to mitigate for the risks this presents.

dezren39 committed:
implement byte-aware chunking

- Add chunkType parameter, allowing multi-byte aware chunking.
   - Third parameter, default behavior if excluded is equal to previous behavior.
   - "-1"  : same as previous default behaviour. every character = 1
   - "0"   : every character = as many bytes as the character is
   - ">0" : every character = as many bytes as the character is, up to N
     - This is the use case I need, as I need each emoji to count as "2".
- Some characters may encode to more bytes than expected.
- Added new tests to accommodate changes and verify backwards compatibility.
  - Made some subjective changes to pre-existing tests to better allow them to be read in a variety of terminals/fonts.
     - When working with emoji, I've had issues where they display differently everywhere.
     - Parsing the variable names has been easier overall, while sometimes more annoying on properly encoding editors.
- Changed chunkSize validation to use Numbers.parseInt and Numbers.isNaN
  - It has better overall compatibility, though does allow weird things like `chunk("text", "2 years", "-1 years")`
- To allow easier compatibility with shell, etc:
  -  npm run-scripts: start, chunk, chunk-text
  -  bin/server.js
  -  package.json bin property
- Added keywords to package.json
- Updated dependencies.
   - Had a hard time parsing between needed dev dependencies.
   - Some _may_ be able to be removed.
- Worked towards ESM-native code with CJS backwards compatibility.
  - Switched code/tests to imports.
  - package.json module, exports, properties.
  - CJS require still compatible.
  - Local build/dev may have new engine compatibility limitations.
  - I was unable to actually set "type":"module", so I clarified package.json property: "type":"commonJs"
    - The issue I had was jest and babel had different compatibility issues, their solutions seemed to conflict.
    - None of the other PRs related to the errors I found had any working solutions.
    - Jest devs didn't get community feedback in 2019 to prioritize ES support.
    - I considered other solutions, but already had made enough major changes for one PR.
       - On this note, I know algolia uses this for a lot of processing or something. I haven't performance tested anything and literally have just seen it "work on my machine". I made the changes for my own needs, but in order to get them mainlined can definitely rip out anything like imports instead of requires or something if it better aligns with something I'm missing. For instance, I updated a bunch of babel things and maybe you had a reason for locking it where it was outside of this module's requirements. 
- Incremented minor version, because although adding a parameter isn't a breaking change --
   - I do worry the overall changes may have some possibility of compatibility issue. Babel helps alleviate this concern though.
   - The effort to allow this capability has increased the complexity of this project.
      - I've tried to add enough testing to mitigate for the risks this presents.

dezren39 committed:
only call "new TextEncoder()" 1 time

dezren39 committed:
add back `package.json` `"main"` property

Haroenv commented on src/index.js:
```diff
@@ -10,36 +10,123 @@ const assertIsValidText = function(text) {
+  const chunkTypeInt = chunkTypeParseIntNaN ? -1 : chunkTypeParseInt;
+  const chunks = [];
+  let characters = runes(text);
+  const textEncoder = new TextEncoder();
```
important to note that adding this is a breaking change, since it changes browser support without a polyfill

dezren39 replied:
😬 understood, hadn't thought of that. I'd figured babel would polyfill it, looking in `dist/`, I see that's not the case.

What do you recommend?

There is the **polyfill in the mdn** link:

https://developer.mozilla.org/en-US/docs/Web/API/TextEncoder

Or the **polyfill mdn suggests for non-utf8**:

https://github.com/inexorabletash/text-encoding

**^ this one** also appears to have an **npm package**, not sure about the mdn polyfill:

https://www.npmjs.com/package/text-encoding

I used **`fast-text-encoding`** for `jest` in dev, but I didn't look at it too closely past for just dev purposes:

https://www.npmjs.com/package/fast-text-encoding

dezren39 replied:
I can also backtrack on using TextEncoder, but it seemed like the best of the options when it is available.
Whatever works best!

dezren39 replied:
Or can call it a breaking change and move on, but I'm unsure your preferences there.

Haroenv replied:
Is there some way to make it optional, only using the TextEncoder, and falling back on a more rough algorithm if it's not given, e.g. `chunk(..., TextEncoder)` (which would allow a polyfill to be passed by the user), where the default one is either the global TextEncoder, or a small alternative with the same API

dezren39 replied:
(working on this)

dezren39 replied:
i've got all this working locally, i have to copy the changes to both branches and still need to fix up some tests. busy with work, hoping this or next weekend i'll have the time

dezren39 committed:
chunkOptions parameter

* bring changes from other repo

* lint:fix

* update test

* textEncoder

* minor update

* cleanup parameter validation

* update 2nd typeexception

* sync with other branch

* add textEncoder dev dependency

dezren39 committed:
update readme

dezren39 committed:
clarify readme

dezren39 committed:
Add TextEncoder to README

dezren39 committed:
Put TextEncoder In README

dezren39 committed:
rephrase

dezren39 committed:
remove fast-text-encoding

dezren39 commented:
OK all set I think. I made an object parameter which I think makes sense in this context. Added a polyfill as a devDependency but not for prod. Readme updated with details. Added some more tests too. I never got it to build on master using my computer, but it builds in the syntax-bin branch for me and the tests pass so I think that it would build for you if the environment is right. I plan to look into what I'm missing at some point.

dezren39 commented:
Hello! Just checking in, is there anything I can to help validate these changes?

Haroenv commented:
sorry @dezren39, looks like this had slipped of our review-list. I'll take a detailed look soon

dezren39 commented:
Thank you! No worries I totally understand!

dezren39 commented:
Howdy! Checking in :-)

Haroenv reviewed (approved):
it makes sense to me 🚀
